core/GlobalSubstitution.cc

```diff
@@ -5,109 +5,49 @@
+            i++;
+            ENFORCE(uniqueNameSubstitution.size() == i, "Unique name substitution has wrong size");
+            if (nm.original.kind() == NameKind::CONSTANT &&
+                nm.original.constantIndex() >= constantNameSubstitution.size()) {
+                // Note: Duplicate of loop body below. If you change one, change the other!
+                for (uint32_t i = constantNameSubstitution.size(); i <= nm.original.constantIndex(); i++) {
+                    auto &cnst = from.constantNames[i];
+                    ENFORCE_NO_TIMER(constantNameSubstitution.size() == i, "Constant name substitution has wrong size");
+                    // N.B.: cnst may reference a UniqueName, but since names are linearizeable we should have
+                    // already substituted it by now.
+                    constantNameSubstitution.emplace_back(to.enterNameConstant(substitute(cnst.original)));
                 }
-
-                uniqueNameSubstitution.emplace_back(
-                    to.freshNameUnique(nm.uniqueNameKind, substitute(nm.original), nm.num));
-                ENFORCE(!fastPath || uniqueNameSubstitution.back().uniqueIndex() == i);
-            }
-            for (i = constantNameSubstitution.size(); i < from.constantNames.size(); i++) {
-                ENFORCE_NO_TIMER(constantNameSubstitution.size() == i, "Constant name substitution has wrong size");
-                auto &nm = from.constantNames[i];
-                constantNameSubstitution.emplace_back(to.enterNameConstant(substitute(nm.original)));
-                ENFORCE(!fastPath || constantNameSubstitution.back().constantIndex() == i);
             }
-        }
 
-        // Enforce that the symbol tables are the same
-        for (int i = 0; i < from.classAndModules.size(); ++i) {
-            ENFORCE(substitute(from.classAndModules[i].name) == from.classAndModules[i].name);
-            ENFORCE(from.classAndModules[i].name == to.classAndModules[i].name);
-        }
-        for (int i = 0; i < from.methods.size(); ++i) {
-            ENFORCE(substitute(from.methods[i].name) == from.methods[i].name);
-            ENFORCE(from.methods[i].name == to.methods[i].name);
-        }
-        for (int i = 0; i < from.fields.size(); ++i) {
-            ENFORCE(substitute(from.fields[i].name) == from.fields[i].name);
-            ENFORCE(from.fields[i].name == to.fields[i].name);
+            uniqueNameSubstitution.emplace_back(to.freshNameUnique(nm.uniqueNameKind, substitute(nm.original), nm.num));
         }
-        for (int i = 0; i < from.typeArguments.size(); ++i) {
-            ENFORCE(substitute(from.typeArguments[i].name) == from.typeArguments[i].name);
-            ENFORCE(from.typeArguments[i].name == to.typeArguments[i].name);
-        }
-        for (int i = 0; i < from.typeMembers.size(); ++i) {
-            ENFORCE(substitute(from.typeMembers[i].name) == from.typeMembers[i].name);
-            ENFORCE(from.typeMembers[i].name == to.typeMembers[i].name);
```

Comment:
it would be nice to figure out a way to make it harder-than-normal to enter symbols into the symbol table, but it's probably fine to not block on that.